framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -2472,40 +2473,53 @@ protected void checkTypecastSafety(TypeCastTree typeCastTree) {
         }
         // We cannot do a simple test of casting, as isSubtypeOf requires
         // the input types to be subtypes according to Java.
-        if (!reported && !isTypeCastSafe(castType, exprType)) {
-            checker.reportWarning(
-                    typeCastTree, "cast.unsafe", exprType.toString(true), castType.toString(true));
+        if (!reported) {
+            TypecastKind castResult = isTypeCastSafe(castType, exprType);
+
+            if (castResult == TypecastKind.WARNING) {
+                checker.reportWarning(
+                        typeCastTree,
+                        "cast.unsafe",
+                        exprType.toString(true),
+                        castType.toString(true));
+            } else if (castResult == TypecastKind.ERROR) {
+                checker.reportError(
+                        typeCastTree,
+                        "cast.incomparable",
+                        exprType.toString(true),
+                        castType.toString(true));
+            }
         }
     }
 
+    /** To represent all possible kinds of typecast */
+    protected enum TypecastKind {
+        /** The cast is safe */
+        SAFE,
+        /** The cast is illegal */
+        ERROR,
+        /** Cannot statically verify the cast, report a warning */
+        WARNING,
+        /** It's not an upcast */
+        NOT_UPCAST,
```

Comment:
Previously, I use `NOT_UPCAST` to represent a state that needs to be further processed, evaluated. `SAFE, ERROR, WARNING` all mean that we already get the result so that we could return our results early.

To remove `NOT_UPCAST`, `NOT_DOWNCAST`, the only way I can think of is putting `null` into the enum variable, to represent current `isTypecastSafe` information is not-yet decided.